var documenterSearchIndex = {"docs":
[{"location":"toy_example/#CpelAsm-1","page":"Toy Example","title":"CpelAsm","text":"","category":"section"},{"location":"toy_example/#Toy-Example-1","page":"Toy Example","title":"Toy Example","text":"","category":"section"},{"location":"toy_example/#","page":"Toy Example","title":"Toy Example","text":"The package includes a small toy example for illustrative purposes. The example consists of two alleles a1 and a2. The former has a mean-methylation level (MML) of 0.8, while that of the later is 0.2. Nevertheless, given the symmetry of the problem, both alleles have the same Shannon entropy. Thus differential  analysis only identifies differences in terms of MML. The output bedGraph files can be found in out_path. To run the toy example run the following commands in a julia's REPL:","category":"page"},{"location":"toy_example/#","page":"Toy Example","title":"Toy Example","text":"using CpelAsm\ndir = \"/path/to/CpelAsm.jl/test/\"\nb1 = \"$(dir)/bam/example.a1.bam\"\nb2 = \"$(dir)/bam/example.a2.bam\"\nfa = \"$(dir)/fasta/n-masked/example.fa\"\nvcf = \"$(dir)/vcf/example.vcf\"\nout = \"$(dir)/out/\"\nrun_analysis(b1,b2,b1,vcf,fa,out;g_max=50,cov_ths=5,cov_b=2.0,win_exp=10,n_null=50,n_max=10)","category":"page"},{"location":"#CpelAsm-1","page":"Home","title":"CpelAsm","text":"","category":"section"},{"location":"#","page":"Home","title":"Home","text":"(Image: CI) (Image: Docs) (Image: MIT license)","category":"page"},{"location":"#Description-1","page":"Home","title":"Description","text":"","category":"section"},{"location":"#","page":"Home","title":"Home","text":"CpelAsm is a julia package especifically desgined for haplotype allele-specific  methylation based on the method in [1]. CpelAsm draws ideas from statistical  physics and information theory to detect allele-specific methylation imbalances  at the haplotype level.","category":"page"},{"location":"#Testing-1","page":"Home","title":"Testing","text":"","category":"section"},{"location":"#","page":"Home","title":"Home","text":"CpelAsm is tested against Julia 1.3.0 on the latest versions of Linux, macOS and Windows.","category":"page"},{"location":"#Installation-1","page":"Home","title":"Installation","text":"","category":"section"},{"location":"#Prerequisites-1","page":"Home","title":"Prerequisites","text":"","category":"section"},{"location":"#","page":"Home","title":"Home","text":"julia v1.3.0\ngit.","category":"page"},{"location":"#Command-1","page":"Home","title":"Command","text":"","category":"section"},{"location":"#","page":"Home","title":"Home","text":"CpelAsm and dependencies can be installed using the following command:","category":"page"},{"location":"#","page":"Home","title":"Home","text":"julia -e 'using Pkg; Pkg.add(PackageSpec(url=\"https://github.com/jordiabante/CpelAsm.jl.git\"))'","category":"page"},{"location":"#Installation-tests-1","page":"Home","title":"Installation tests","text":"","category":"section"},{"location":"#","page":"Home","title":"Home","text":"In a julia session run","category":"page"},{"location":"#","page":"Home","title":"Home","text":"(v1.3) pkg> test CpelAsm","category":"page"},{"location":"#Authors-1","page":"Home","title":"Authors","text":"","category":"section"},{"location":"#","page":"Home","title":"Home","text":"Jordi Abante","category":"page"},{"location":"#References-1","page":"Home","title":"References","text":"","category":"section"},{"location":"#","page":"Home","title":"Home","text":"[1] Abante, J., Fang, Y., Feinberg, A.P., Goutsias, J., Detection of haplotype-dependent  allele-speciÔ¨Åc DNA methylation in WGBS data, Nature Communications 2020 XYZ.","category":"page"}]
}
