<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Main Commands · CpelAsm</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../"><img src="../assets/logo.png" alt="CpelAsm logo"/></a><div class="docs-package-name"><span class="docs-autofit">CpelAsm</span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li class="is-active"><a class="tocitem" href>Main Commands</a><ul class="internal"><li><a class="tocitem" href="#Generate-GFF-files-1"><span>Generate GFF files</span></a></li><li><a class="tocitem" href="#BedGraphs-MML,-NME,-and-PDM-1"><span>BedGraphs MML, NME, &amp; PDM</span></a></li><li><a class="tocitem" href="#Generate-Null-Statistics-1"><span>Generate Null Statistics</span></a></li><li><a class="tocitem" href="#Perform-Allele-Specific-Methylation-Detection-1"><span>Perform Allele-Specific Methylation Detection</span></a></li><li><a class="tocitem" href="#Running-CpelAsm-with-a-single-command-1"><span>Running CpelAsm with a single command</span></a></li></ul></li><li><a class="tocitem" href="../toy_example/">Toy Example</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Main Commands</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Main Commands</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/jordiabante/CpelAsm.jl/blob/master/docs/src/main_commands.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="CpelAsm-1"><a class="docs-heading-anchor" href="#CpelAsm-1">CpelAsm</a><a class="docs-heading-anchor-permalink" href="#CpelAsm-1" title="Permalink"></a></h1><p><img src="https://github.com/jordiabante/CpelAsm.jl/workflows/CI/badge.svg" alt="CI"/> <img src="https://github.com/jordiabante/CpelAsm.jl/workflows/Docs/badge.svg" alt="Docs"/> <a href="https://github.com/jordiabante/CpelAsm.jl/blob/master/LICENSE.md"><img src="https://img.shields.io/badge/license-MIT-green.svg" alt="MIT license"/></a></p><h2 id="Generate-GFF-files-1"><a class="docs-heading-anchor" href="#Generate-GFF-files-1">Generate GFF files</a><a class="docs-heading-anchor-permalink" href="#Generate-GFF-files-1" title="Permalink"></a></h2><p>The first step consists in generating the 2 necessary GFF files: heterozygous and homozygous. The former is the one that contains the haplotypes analyzed, while the latter contains the homozygous regions of the genome.</p><p>In the following example, the maximum number of CpG sites allowed per haplotype is 25 (<code>n_max=25</code>).  In addition, CpelAsm extends the window delimited by the first and last SNP in the haplotype by 100  bp left and right (<code>win_exp=100</code>).</p><pre><code class="language-julia"># Parameters
n_max = 25
win_exp = 100

# Paths
project_path = &quot;/path/to/SampleName&quot;
vcf = &quot;$(project_path)/vcf/SampleName.phased.vcf&quot;
fasta = &quot;$(project_path)/fasta/SampleName.masked_hg19.fa&quot;
het_gff = &quot;$(project_path)/CpelAsmOut/SampleName_het.cpelasm.gff&quot;
hom_gff = &quot;$(project_path)/CpelAsmOut/SampleName_hom.cpelasm.gff&quot;

# Call
gen_gffs([het_gff,hom_gff],fasta,vcf,win_exp,n_max)
</code></pre><h2 id="BedGraphs-MML,-NME,-and-PDM-1"><a class="docs-heading-anchor" href="#BedGraphs-MML,-NME,-and-PDM-1">BedGraphs MML, NME, &amp; PDM</a><a class="docs-heading-anchor-permalink" href="#BedGraphs-MML,-NME,-and-PDM-1" title="Permalink"></a></h2><p>The next step consists in estimating the allele-specific Ising models in the haplotypes and generating bedGraph files with MML1/2, NME1/2, and PDM. CpelAsm can be parallelized  when multiple CPUs are available by loading first the package <code>Distributed</code> and then  loading CpelAsm through the macro `<code>@everywhere</code>.</p><p>In the following example, the maximum size of a subregion is fixed to 500 (<code>g_max=500</code>),  the minimum average depth is set to 5 (<code>cov_ths=5</code>), and the WGBS reads are trimmed 5 bp on each end (<code>trim=(5,5,5,5)</code>).</p><pre><code class="language-julia"># Deps
using Distributed
@everywhere using CpelAsm

# Parameters
g_max = 500
cov_ths = 5
trim = (5,5,5,5)

# Paths
project_path = &quot;/path/to/SampleName&quot;
fasta = &quot;$(project_path)/fasta/SampleName.masked_hg19.fa&quot;
het_gff = &quot;$(project_path)/cpelasm/SampleName_het.cpelasm.gff&quot;
bam1 = &quot;$(project_path)/bam/SampleName.sort.genome1.bam&quot;
bam2 = &quot;$(project_path)/bam/SampleName.sort.genome2.bam&quot;
pdm_path = &quot;$(project_path)/cpelasm/SampleName_pdm.bedGraph&quot;
mml1_path = &quot;$(project_path)/cpelasm/SampleName_mml1.bedGraph&quot;
mml2_path = &quot;$(project_path)/cpelasm/SampleName_mml2.bedGraph&quot;
nme1_path = &quot;$(project_path)/cpelasm/SampleName_nme1.bedGraph&quot;
nme2_path = &quot;$(project_path)/cpelasm/SampleName_nme2.bedGraph&quot;
tobs_path = [mml1_path,mml2_path,nme1_path,nme2_path,pdm_path]

# Call
comp_tobs(bam1,bam2,het_gff,fasta,tobs_path;g_max=g_max,cov_ths=cov_ths,trim=trim)
</code></pre><h2 id="Generate-Null-Statistics-1"><a class="docs-heading-anchor" href="#Generate-Null-Statistics-1">Generate Null Statistics</a><a class="docs-heading-anchor-permalink" href="#Generate-Null-Statistics-1" title="Permalink"></a></h2><p>The next step consists in generating null statistics to be able to perform hypothesis testing. As discussed in the previous point, CpelAsm can be parallelized when multiple  CPUs are available by first loading the package <code>Distributed</code> and then loading CpelAsm  through the macro `<code>@everywhere</code>.</p><p>In the following example, the maximum size of a subregion is fixed to 500 (<code>g_max=500</code>),  the minimum average depth is set to 5 (<code>cov_ths=5</code>), and the WGBS reads are trimmed 5 bp on each end (<code>trim=(5,5,5,5)</code>).</p><pre><code class="language-julia"># Deps
using Distributed
@everywhere using CpelAsm

# Parameters
cov_ths = 5
g_max = 500
trim = (5,5,5,5)

# Paths
project_path = &quot;/path/to/SampleName&quot;
fasta = &quot;$(project_path)/fasta/SampleName.masked_hg19.fa&quot;
bam1 = &quot;$(project_path)/bam/SampleName.sort.genome1.bam&quot;
bam2 = &quot;$(project_path)/bam/SampleName.sort.genome2.bam&quot;
het_gff = &quot;$(project_path)/cpelasm/SampleName_het.cpelasm.gff&quot;
hom_gff = &quot;$(project_path)/cpelasm/SampleName_hom.cpelasm.gff&quot;
null_pdm_path = &quot;$(project_path)/cpelasm/SampleName_pdm_null.bedGraph&quot;
null_tmml_path = &quot;$(project_path)/cpelasm/SampleName_tmml_null.bedGraph&quot;
null_tnme_path = &quot;$(project_path)/cpelasm/SampleName_tnme_null.bedGraph&quot;
tnull_path = [null_tmml_path,null_tnme_path,null_pdm_path]
pdm_path = &quot;$(project_path)/cpelasm/SampleName_pdm.bedGraph&quot;
mml1_path = &quot;$(project_path)/cpelasm/SampleName_mml1.bedGraph&quot;
mml2_path = &quot;$(project_path)/cpelasm/SampleName_mml2.bedGraph&quot;
nme1_path = &quot;$(project_path)/cpelasm/SampleName_nme1.bedGraph&quot;
nme2_path = &quot;$(project_path)/cpelasm/SampleName_nme2.bedGraph&quot;
tobs_path = [mml1_path,mml2_path,nme1_path,nme2_path,pdm_path]

# Call
comp_tnull(bam,het_gff,hom_gff,fasta,tobs_path,tnull_path;
    g_max=g_max,cov_ths=cov_ths,trim=trim,n_max=n_max)
</code></pre><h2 id="Perform-Allele-Specific-Methylation-Detection-1"><a class="docs-heading-anchor" href="#Perform-Allele-Specific-Methylation-Detection-1">Perform Allele-Specific Methylation Detection</a><a class="docs-heading-anchor-permalink" href="#Perform-Allele-Specific-Methylation-Detection-1" title="Permalink"></a></h2><p>The final step is to compute a p-value for each haplotype for each one of the three quantities (Tmml, Tnme, Tpdm). The returned p-values are corrected using the  Benjamini-Hochberg and allow for control of the false discovery rate (FDR).</p><p>In the following example, hypothesis testing is performed for haplotypes with at  most 25 CpG sites (<code>n_max=25</code>), and a minimum of 1,000 null statistics to perform hypothesis testing.</p><pre><code class="language-julia"># Deps
using CpelAsm

# Parameters
n_max = 25
n_null = 1000

# Paths
project_path = &quot;/path/to/SampleName&quot;
fasta = &quot;$(project_path)/fasta/SampleName.masked_hg19.fa&quot;
bam1 = &quot;$(project_path)/bam/SampleName.sort.genome1.bam&quot;
bam2 = &quot;$(project_path)/bam/SampleName.sort.genome2.bam&quot;
null_pdm_path = &quot;$(project_path)/cpelasm/SampleName_pdm_null.bedGraph&quot;
null_tmml_path = &quot;$(project_path)/cpelasm/SampleName_tmml_null.bedGraph&quot;
null_tnme_path = &quot;$(project_path)/cpelasm/SampleName_tnme_null.bedGraph&quot;
tnull_path = [null_tmml_path,null_tnme_path,null_pdm_path]
pdm_path = &quot;$(project_path)/cpelasm/SampleName_pdm.bedGraph&quot;
mml1_path = &quot;$(project_path)/cpelasm/SampleName_mml1.bedGraph&quot;
mml2_path = &quot;$(project_path)/cpelasm/SampleName_mml2.bedGraph&quot;
nme1_path = &quot;$(project_path)/cpelasm/SampleName_nme1.bedGraph&quot;
nme2_path = &quot;$(project_path)/cpelasm/SampleName_nme2.bedGraph&quot;
tobs_path = [mml1_path,mml2_path,nme1_path,nme2_path,pdm_path]
pVal_tmml_path = &quot;$(project_path)/cpelasm/SampleName_tmml_pvals.bedGraph&quot;
pVal_tnme_path = &quot;$(project_path)/cpelasm/SampleName_tnme_pvals.bedGraph&quot;
pVal_tpdm_path = &quot;$(project_path)/cpelasm/SampleName_tpdm_pvals.bedGraph&quot;
pVal_path = [pVal_tmml_path,pVal_tnme_path,pVal_tpdm_path]

# Call
comp_pvals(tobs_path,tnull_path,p_path,n_max,n_null)
</code></pre><h2 id="Running-CpelAsm-with-a-single-command-1"><a class="docs-heading-anchor" href="#Running-CpelAsm-with-a-single-command-1">Running CpelAsm with a single command</a><a class="docs-heading-anchor-permalink" href="#Running-CpelAsm-with-a-single-command-1" title="Permalink"></a></h2><p>The following example shows how to perform all the steps shown above in a single command.</p><pre><code class="language-julia"># Deps
using Distributed
@everywhere using CpelAsm

# Parameters
cov_ths = 5
g_max = 500
win_exp = 100
trim = (5,5,5,5)

# Paths
project_path = &quot;/path/to/SampleName&quot;
vcf = &quot;$(project_path)/vcf/SampleName.phased.vcf&quot;
fasta = &quot;$(project_path)/fasta/SampleName.masked_hg19.fa&quot;
bam1 = &quot;$(project_path)/bam/SampleName.sort.genome1.bam&quot;
bam2 = &quot;$(project_path)/bam/SampleName.sort.genome2.bam&quot;
bamu = &quot;$(project_path)/bam/SampleName.sort.unassigned.bam&quot;
outdir = &quot;$(project_path)/cpelasm/&quot;

# Call
run_analysis(bam1,bam2,bamu,vcf,fasta,outdir;g_max=g_max,cov_ths=cov_ths,
    win_exp=win_exp,trim=trim,n_null=n_null,n_max=n_max)
</code></pre></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../">« Home</a><a class="docs-footer-nextpage" href="../toy_example/">Toy Example »</a></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Wednesday 22 April 2020 16:03">Wednesday 22 April 2020</span>. Using Julia version 1.3.1.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
